File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ConnectionPause.kt
package expo.modules.vescapecore.connection

import android.content.Context
import expo.modules.vescapecore.diagnostics.ConnectionTraceEvent
import expo.modules.vescapecore.diagnostics.ConnectionTraceField
import expo.modules.vescapecore.diagnostics.ConnectionTraceReason
import expo.modules.vescapecore.diagnostics.ConnectionWorkflow
import org.json.JSONObject

/**
 * One Automatic Connection Pause entry: a Board the rider deliberately stopped, and the absolute
 * moment automatic connection may resume for it (ADR 0035).
 *
 * @parity /modules/vescape-core/ios/connection/ConnectionPause.swift `ConnectionPause`
 * @parity /modules/vescape-core/src/index.ts `ConnectionPauseState`
 */
data class ConnectionPause(
    val boardId: String,
    /** Absolute deadline. Expiry is by clock comparison, so no cleanup job exists. */
    val untilMs: Long,
    /** Rider action that armed it, from [ConnectionTraceReason]. */
    val source: String,
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "boardId" to boardId,
        "until" to untilMs,
        "source" to source,
    )
}

/**
 * Pure rules of the board-scoped Automatic Connection Pause map (ADR 0035, #406). This replaces the
 * permanent manual-stop tombstone and the separate Auto Start restart gate: one map, one deadline
 * per Board, shared by Auto Connect and Android Auto Start.
 *
 * Rider intent arms a pause; mechanics never do. Mechanical teardown, Board switch cleanup, probe
 * cancellation, Stop search, and scan timeout are deliberately absent from [ARMING_SOURCES].
 *
 * @parity /modules/vescape-core/ios/connection/ConnectionPause.swift `ConnectionPausePolicy`
 */
object ConnectionPausePolicy {
    /** Legacy values up to 24h stay valid; only the rider-facing stepper recommends less. */
    const val MAX_PAUSE_MINUTES = 1440

    /** Cap offered for *new* selections. Stored values above it are preserved, never clamped. */
    const val RECOMMENDED_MAX_PAUSE_MINUTES = 480

    /** The four rider actions that arm a pause. Everything else is mechanics. */
    val ARMING_SOURCES: Set<String> = setOf(
        ConnectionTraceReason.MANUAL_DISCONNECT,
        ConnectionTraceReason.END_RIDE,
        ConnectionTraceReason.APP_EXIT,
        ConnectionTraceReason.TASK_REMOVED,
    )

    fun arms(source: String): Boolean = source in ARMING_SOURCES

    /** Absolute deadline for a pause armed now, or `null` when the rider configured zero minutes. */
    fun deadlineFor(nowMs: Long, minutes: Int): Long? =
        if (minutes <= 0) null else nowMs + minutes * 60_000L

    fun isActive(pause: ConnectionPause, nowMs: Long): Boolean = pause.untilMs > nowMs

    /** The still-running pause for [boardId], or `null` when it never existed or already expired. */
    fun active(entries: Map<String, ConnectionPause>, boardId: String?, nowMs: Long): ConnectionPause? {
        if (boardId.isNullOrBlank()) return null
        return entries[boardId]?.takeIf { isActive(it, nowMs) }
    }

    /** Drop expired entries. Returns the same instance when nothing expired. */
    fun prune(entries: Map<String, ConnectionPause>, nowMs: Long): Map<String, ConnectionPause> {
        val kept = entries.filterValues { isActive(it, nowMs) }
        return if (kept.size == entries.size) entries else kept
    }
}

/**
 * Where the Automatic Connection Pause map is kept. The one Android-typed line of the feature, so
 * the map's arm/clear/expire semantics stay unit-testable without a device.
 *
 * iOS reaches the same seam by injecting `UserDefaults` into `ConnectionPauseStore`.
 */
internal interface ConnectionPauseStorage {
    fun load(): String?

    fun save(value: String?)
}

/**
 * The Automatic Connection Pause map itself: arm, clear, and read entries by Board id, persisted as
 * JSON through [ConnectionPauseStorage]. Expiry happens on read, so no cleanup job exists.
 *
 * @parity /modules/vescape-core/ios/connection/ConnectionPause.swift `ConnectionPauseStore`
 */
internal class ConnectionPauseRegistry(
    private val storage: ConnectionPauseStorage,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    /**
     * Arm a pause for [boardId]. Returns `null` — and stores nothing — when [source] is not a rider
     * action, or when the configured duration is zero (the rider opted out of pausing).
     */
    @Synchronized
    fun arm(
        boardId: String?,
        source: String,
        minutes: Int,
        workflow: ConnectionWorkflow? = null,
    ): ConnectionPause? {
        if (boardId.isNullOrBlank() || !ConnectionPausePolicy.arms(source)) return null
        val now = clock()
        val until = ConnectionPausePolicy.deadlineFor(now, minutes) ?: run {
            workflow?.event(
                ConnectionTraceEvent.PAUSE_BLOCKED,
                mapOf(
                    ConnectionTraceField.BOARD_ID to boardId,
                    ConnectionTraceField.PAUSE_SOURCE to source,
                    ConnectionTraceField.DEADLINE_MS to 0L,
                ),
            )
            return null
        }
        val pause = ConnectionPause(boardId = boardId, untilMs = until, source = source)
        write(ConnectionPausePolicy.prune(read(), now) + (boardId to pause))
        workflow?.event(
            ConnectionTraceEvent.PAUSE_STARTED,
            mapOf(
                ConnectionTraceField.BOARD_ID to boardId,
                ConnectionTraceField.PAUSE_SOURCE to source,
                ConnectionTraceField.PAUSED_UNTIL to until,
            ),
        )
        return pause
    }

    /** Explicit Connect, Connect now, and Switch & Connect clear the affected Board's pause. */
    @Synchronized
    fun clear(boardId: String?, workflow: ConnectionWorkflow? = null) {
        if (boardId.isNullOrBlank()) return
        val entries = read()
        val existing = entries[boardId] ?: return
        write(entries - boardId)
        workflow?.event(
            ConnectionTraceEvent.PAUSE_CLEARED,
            mapOf(
                ConnectionTraceField.BOARD_ID to boardId,
                ConnectionTraceField.PAUSE_SOURCE to existing.source,
                ConnectionTraceField.PAUSED_UNTIL to existing.untilMs,
            ),
        )
    }

    /** The running pause for [boardId]. Expired entries are dropped here — no cleanup job. */
    @Synchronized
    fun active(boardId: String?, workflow: ConnectionWorkflow? = null): ConnectionPause? {
        if (boardId.isNullOrBlank()) return null
        val now = clock()
        val entries = read()
        val stored = entries[boardId] ?: return null
        if (ConnectionPausePolicy.isActive(stored, now)) return stored
        write(ConnectionPausePolicy.prune(entries, now))
        workflow?.event(
            ConnectionTraceEvent.PAUSE_EXPIRED,
            mapOf(
                ConnectionTraceField.BOARD_ID to boardId,
                ConnectionTraceField.PAUSE_SOURCE to stored.source,
                ConnectionTraceField.PAUSED_UNTIL to stored.untilMs,
            ),
        )
        return null
    }

    /** Deadline feeding [PresencePromotionInput.pausedUntilMs], or `null` when not paused. */
    fun pausedUntilMs(boardId: String?, workflow: ConnectionWorkflow? = null): Long? =
        active(boardId, workflow)?.untilMs

    @Synchronized
    fun entries(): Map<String, ConnectionPause> = read()

    @Synchronized
    fun clearAll() {
        storage.save(null)
    }

    private fun read(): Map<String, ConnectionPause> {
        val raw = storage.load() ?: return emptyMap()
        return try {
            val json = JSONObject(raw)
            buildMap {
                for (boardId in json.keys()) {
                    val entry = json.optJSONObject(boardId) ?: continue
                    val until = entry.optLong(FIELD_UNTIL, 0L)
                    val source = entry.optString(FIELD_SOURCE).takeIf { it.isNotBlank() } ?: continue
                    if (until <= 0L) continue
                    put(boardId, ConnectionPause(boardId, until, source))
                }
            }
        } catch (_: Exception) {
            // A corrupt map must not permanently block automatic connection: fail open, not shut.
            emptyMap()
        }
    }

    private fun write(entries: Map<String, ConnectionPause>) {
        if (entries.isEmpty()) {
            storage.save(null)
            return
        }
        val json = JSONObject()
        for ((boardId, pause) in entries) {
            json.put(
                boardId,
                JSONObject().put(FIELD_UNTIL, pause.untilMs).put(FIELD_SOURCE, pause.source),
            )
        }
        storage.save(json.toString())
    }

    private companion object {
        const val FIELD_UNTIL = "until"
        const val FIELD_SOURCE = "source"
    }
}

/**
 * Process-wide Automatic Connection Pause map, persisted in SharedPreferences. It survives process
 * death and reboot, because the whole point is that force-quitting the app does not re-arm automatic
 * connection.
 *
 * Every entry point takes a Board id explicitly, never "the selected Board": Android Auto Start
 * (#407) evaluates the *detected* Board, and Switch & Connect (#408) clears the *target* Board.
 *
 * @parity /modules/vescape-core/ios/connection/ConnectionPause.swift `ConnectionPauseStore`
 */
object ConnectionPauseStore {
    const val PREFS = "vesc_automatic_connection_pause"
    private const val KEY_ENTRIES = "entries"

    fun arm(
        context: Context,
        boardId: String?,
        source: String,
        minutes: Int,
        workflow: ConnectionWorkflow? = null,
    ): ConnectionPause? = registry(context).arm(boardId, source, minutes, workflow)

    fun clear(context: Context, boardId: String?, workflow: ConnectionWorkflow? = null) =
        registry(context).clear(boardId, workflow)

    fun active(
        context: Context,
        boardId: String?,
        workflow: ConnectionWorkflow? = null,
    ): ConnectionPause? = registry(context).active(boardId, workflow)

    fun pausedUntilMs(
        context: Context,
        boardId: String?,
        workflow: ConnectionWorkflow? = null,
    ): Long? = registry(context).pausedUntilMs(boardId, workflow)

    fun entries(context: Context): Map<String, ConnectionPause> = registry(context).entries()

    private fun registry(context: Context) = ConnectionPauseRegistry(
        object : ConnectionPauseStorage {
            private val prefs =
                context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

            override fun load(): String? = prefs.getString(KEY_ENTRIES, null)

            override fun save(value: String?) {
                val editor = prefs.edit()
                if (value == null) editor.remove(KEY_ENTRIES) else editor.putString(KEY_ENTRIES, value)
                editor.apply()
            }
        },
    )
}

File: modules/vescape-core/ios/connection/ConnectionPause.swift
import Foundation

/// One Automatic Connection Pause entry: a Board the rider deliberately stopped, and the absolute
/// moment automatic connection may resume for it (ADR 0035).
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ConnectionPause.kt `ConnectionPause`
/// @parity /modules/vescape-core/src/index.ts `ConnectionPauseState`
struct ConnectionPause: Equatable {
  let boardId: String
  /// Absolute deadline. Expiry is by clock comparison, so no cleanup job exists.
  let untilMs: Int64
  /// Rider action that armed it, from `ConnectionTraceReason`.
  let source: String

  var map: [String: Any?] {
    ["boardId": boardId, "until": untilMs, "source": source]
  }
}

/// Pure rules of the board-scoped Automatic Connection Pause map (ADR 0035, #406). This replaces the
/// permanent manual-stop tombstone and the separate Auto Start restart gate: one map, one deadline
/// per Board, shared by Auto Connect and Android Auto Start.
///
/// Rider intent arms a pause; mechanics never do. Mechanical teardown, Board switch cleanup, probe
/// cancellation, Stop search, and scan timeout are deliberately absent from `armingSources`.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ConnectionPause.kt `ConnectionPausePolicy`
enum ConnectionPausePolicy {
  /// Legacy values up to 24h stay valid; only the rider-facing stepper recommends less.
  static let maxPauseMinutes = 1440

  /// Cap offered for *new* selections. Stored values above it are preserved, never clamped.
  static let recommendedMaxPauseMinutes = 480

  /// The four rider actions that arm a pause. Everything else is mechanics.
  static let armingSources: Set<String> = [
    ConnectionTraceReason.manualDisconnect,
    ConnectionTraceReason.endRide,
    ConnectionTraceReason.appExit,
    ConnectionTraceReason.taskRemoved,
  ]

  static func arms(_ source: String) -> Bool { armingSources.contains(source) }

  /// Absolute deadline for a pause armed now, or `nil` when the rider configured zero minutes.
  static func deadlineFor(nowMs: Int64, minutes: Int) -> Int64? {
    minutes <= 0 ? nil : nowMs + Int64(minutes) * 60_000
  }

  static func isActive(_ pause: ConnectionPause, nowMs: Int64) -> Bool { pause.untilMs > nowMs }

  /// The still-running pause for `boardId`, or `nil` when it never existed or already expired.
  static func active(
    entries: [String: ConnectionPause],
    boardId: String?,
    nowMs: Int64
  ) -> ConnectionPause? {
    guard let boardId, !boardId.isEmpty, let stored = entries[boardId] else { return nil }
    return isActive(stored, nowMs: nowMs) ? stored : nil
  }

  /// Drop expired entries.
  static func prune(entries: [String: ConnectionPause], nowMs: Int64) -> [String: ConnectionPause] {
    entries.filter { isActive($0.value, nowMs: nowMs) }
  }
}

/// Persistence for the Automatic Connection Pause map. Survives process death and reboot, because
/// the whole point is that force-quitting the app does not re-arm automatic connection.
///
/// Later slices call it by Board id, never by "the selected Board": Android Auto Start (#407)
/// evaluates the *detected* Board, and Switch & Connect (#408) clears the *target* Board.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/connection/ConnectionPause.kt `ConnectionPauseStore`
final class ConnectionPauseStore {
  static let storageKey = "vesc_automatic_connection_pause"
  static let shared = ConnectionPauseStore()

  private let defaults: UserDefaults
  private let lock = NSLock()

  init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  /// Arm a pause for `boardId`. Returns `nil` — and stores nothing — when `source` is not a rider
  /// action, or when the configured duration is zero (the rider opted out of pausing).
  @discardableResult
  func arm(
    boardId: String?,
    source: String,
    minutes: Int,
    workflow: ConnectionWorkflow? = nil,
    nowMs: Int64 = ConnectionTrace.now()
  ) -> ConnectionPause? {
    guard let boardId, !boardId.isEmpty, ConnectionPausePolicy.arms(source) else { return nil }
    guard let until = ConnectionPausePolicy.deadlineFor(nowMs: nowMs, minutes: minutes) else {
      workflow?.event(
        ConnectionTraceEvent.pauseBlocked,
        fields: [
          ConnectionTraceField.boardId: boardId,
          ConnectionTraceField.pauseSource: source,
          ConnectionTraceField.deadlineMs: 0,
        ]
      )
      return nil
    }
    let pause = ConnectionPause(boardId: boardId, untilMs: until, source: source)
    lock.lock()
    var entries = ConnectionPausePolicy.prune(entries: read(), nowMs: nowMs)
    entries[boardId] = pause
    write(entries)
    lock.unlock()
    workflow?.event(
      ConnectionTraceEvent.pauseStarted,
      fields: [
        ConnectionTraceField.boardId: boardId,
        ConnectionTraceField.pauseSource: source,
        ConnectionTraceField.pausedUntil: until,
      ]
    )
    return pause
  }

  /// Explicit Connect, Connect now, and Switch & Connect clear the affected Board's pause.
  func clear(boardId: String?, workflow: ConnectionWorkflow? = nil) {
    guard let boardId, !boardId.isEmpty else { return }
    lock.lock()
    var entries = read()
    guard let existing = entries[boardId] else {
      lock.unlock()
      return
    }
    entries.removeValue(forKey: boardId)
    write(entries)
    lock.unlock()
    workflow?.event(
      ConnectionTraceEvent.pauseCleared,
      fields: [
        ConnectionTraceField.boardId: boardId,
        ConnectionTraceField.pauseSource: existing.source,
        ConnectionTraceField.pausedUntil: existing.untilMs,
      ]
    )
  }

  /// The running pause for `boardId`. Expired entries are dropped here — no cleanup job.
  func active(
    boardId: String?,
    workflow: ConnectionWorkflow? = nil,
    nowMs: Int64 = ConnectionTrace.now()
  ) -> ConnectionPause? {
    guard let boardId, !boardId.isEmpty else { return nil }
    lock.lock()
    let entries = read()
    guard let stored = entries[boardId] else {
      lock.unlock()
      return nil
    }
    if ConnectionPausePolicy.isActive(stored, nowMs: nowMs) {
      lock.unlock()
      return stored
    }
    write(ConnectionPausePolicy.prune(entries: entries, nowMs: nowMs))
    lock.unlock()
    workflow?.event(
      ConnectionTraceEvent.pauseExpired,
      fields: [
        ConnectionTraceField.boardId: boardId,
        ConnectionTraceField.pauseSource: stored.source,
        ConnectionTraceField.pausedUntil: stored.untilMs,
      ]
    )
    return nil
  }

  /// Deadline feeding `PresencePromotionInput.pausedUntilMs`, or `nil` when not paused.
  func pausedUntilMs(
    boardId: String?,
    workflow: ConnectionWorkflow? = nil,
    nowMs: Int64 = ConnectionTrace.now()
  ) -> Int64? {
    active(boardId: boardId, workflow: workflow, nowMs: nowMs)?.untilMs
  }

  func entries() -> [String: ConnectionPause] {
    lock.lock()
    defer { lock.unlock() }
    return read()
  }

  func clearAll() {
    lock.lock()
    defaults.removeObject(forKey: Self.storageKey)
    lock.unlock()
  }

  // MARK: - Storage

  private func read() -> [String: ConnectionPause] {
    guard let raw = defaults.dictionary(forKey: Self.storageKey) else { return [:] }
    var entries: [String: ConnectionPause] = [:]
    for (boardId, value) in raw {
      guard
        let entry = value as? [String: Any],
        let until = (entry["until"] as? NSNumber)?.int64Value, until > 0,
        let source = entry["source"] as? String, !source.isEmpty
      else { continue }
      entries[boardId] = ConnectionPause(boardId: boardId, untilMs: until, source: source)
    }
    return entries
  }

  private func write(_ entries: [String: ConnectionPause]) {
    guard !entries.isEmpty else {
      defaults.removeObject(forKey: Self.storageKey)
      return
    }
    var raw: [String: Any] = [:]
    for (boardId, pause) in entries {
      raw[boardId] = ["until": NSNumber(value: pause.untilMs), "source": pause.source]
    }
    defaults.set(raw, forKey: Self.storageKey)
  }
}
